### PR TITLE
Add unit tests for LoyaltyCard class

### PR DESCRIPTION
**Pull request by Aglag257**
- Added LoyaltyCardTest to test the parcelable implementation of the LoyaltyCard class
- Implemented tests for the following scenarios:
  - testParcelable: Verifies that LoyaltyCard objects can be correctly serialized and deserialized using Parcel
  - testIsDuplicate_sameObject: Checks that isDuplicate correctly identifies the same object as a duplicate
  - testIsDuplicate_differentObjects: Checks that isDuplicate correctly identifies different objects as not duplicates
  - testToString: Validates the toString method output of LoyaltyCard

This improves test coverage and ensures the correct behavior of the LoyaltyCard class.

### CODE REVIEW

File: app/src/test/java/protect/card_locker/LoyaltyCardTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import java.math.BigDecimal;
+import java.util.Currency;
+import java.util.Date;
+
+import com.google.zxing.BarcodeFormat;
+
+import static org.junit.Assert.*;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(manifest=Config.NONE)
```
Comment (TheLastProject) [inline app/src/test/java/protect/card_locker/LoyaltyCardTest.java:19]: Just curious, why this line? No other tests seem to use it.

File: app/src/test/java/protect/card_locker/LoyaltyCardTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                "LoyaltyCard{%n  id=%s,%n  store=%s,%n  note=%s,%n  validFrom=%s,%n  expiry=%s,%n"
+                        + "  balance=%s,%n  balanceType=%s,%n  cardId=%s,%n  barcodeId=%s,%n  barcodeType=%s,%n"
+                        + "  headerColor=%s,%n  starStatus=%s,%n  lastUsed=%s,%n  zoomLevel=%s,%n  archiveStatus=%s%n}",
+                card.id, card.store, card.note, card.validFrom, card.expiry,
+                card.balance, card.balanceType, card.cardId, card.barcodeId,
+                card.barcodeType != null ? card.barcodeType.format() : null,
+                card.headerColor, card.starStatus, card.lastUsed, card.zoomLevel, card.archiveStatus);
+
+        assertEquals(expected, card.toString());
+    }
```
Comment (TheLastProject) [inline app/src/test/java/protect/card_locker/LoyaltyCardTest.java:95]: I don't think this test is really useful, I don't think the toString is ever used except to be able to use in quick debugging.

Although I guess not testing this would make test coverage tooling say this code isn't tested (even though I don't think it's important to test).

Eh, no strong opinion I guess. Do you have an opinion @obfusk?

**Review by TheLastProject — COMMENTED** [review]
Seems like a logical thing to test. Got just one question :)

File: app/src/test/java/protect/card_locker/LoyaltyCardTest.java (hunk 1 above)
Comment (obfusk) [reply app/src/test/java/protect/card_locker/LoyaltyCardTest.java:19]: Wondering the same. It also seems to be [deprecated](https://robolectric.org/javadoc/latest/org/robolectric/annotation/Config.html#manifest()).

File: app/src/test/java/protect/card_locker/LoyaltyCardTest.java (hunk 2 above)
Comment (obfusk) [reply app/src/test/java/protect/card_locker/LoyaltyCardTest.java:95]: I have to agree it doesn't really serve much purpose, especially as it currently duplicates the exact same `String.format()` from the `.toString()` code.

If you really want a regression test for code like this I would recommend comparing against a hardcoded string of the *expected output* so you can at least confirm the output looks like it's supposed to and aren't essentially just checking whether two identical calls to `String.format()` produce the same result.

I'd say more test coverage is generally good as long as it doesn't require effort to write and maintain the tests that could better be spent on more important things :)

**Comment by obfusk** [comment]
Given that `isDuplicate()` should ignore differences in lastUsed & zoomLevel it seems like a good idea to test that as well.

File: app/src/test/java/protect/card_locker/LoyaltyCardTest.java (hunk 2 above)
Comment (TheLastProject) [reply app/src/test/java/protect/card_locker/LoyaltyCardTest.java:95]: That would be a good change, yeah :+1:

File: app/src/test/java/protect/card_locker/LoyaltyCardTest.java (hunk 1 above)
Comment (Aglag257) [reply app/src/test/java/protect/card_locker/LoyaltyCardTest.java:19]: Thank you for bringing this to my attention.

The @Config(manifest=Config.NONE) line was included to ensure that Robolectric tests run without relying on an Android manifest file. However, I see that the manifest field within the @Config annotation is marked as deprecated. This means that we should migrate to the preferred way of configuring builds as recommended in the Robolectric documentation.

I will update the tests to remove the deprecated usage and follow the new guidelines

File: app/src/test/java/protect/card_locker/LoyaltyCardTest.java (hunk 2 above)
Comment (Aglag257) [reply app/src/test/java/protect/card_locker/LoyaltyCardTest.java:93]: Thank you both for your feedback and suggestions.

I understand your point about the current implementation of the testToString method. I initially included it to have a more complete test class but It indeed duplicates the String.format logic, which may not add significant value to the test coverage. To address this, I can update the test to compare the toString output against a hardcoded expected string. This will help ensure that the output is as expected without duplicating the code logic.

**Comment by Aglag257** [comment]
> Given that `isDuplicate()` should ignore differences in lastUsed & zoomLevel it seems like a good idea to test that as well.

You're right; given that isDuplicate() is designed to ignore differences in lastUsed and zoomLevel, it makes sense to include tests that verify this behavior.

I will add additional tests to ensure that isDuplicate correctly identifies duplicate LoyaltyCard instances even when lastUsed and zoomLevel values differ.
